code/parachain/frame/liquid-staking/src/lib.rs

```diff
@@ -0,0 +1,2105 @@
+				derivative_index,
+				key,
+				derivative_account_id,
+			));
+			Ok(())
+		}
+
+		#[pallet::call_index(26)]
+		#[pallet::weight(<T as Config>::WeightInfo::update_incentive())]
+		#[transactional]
+		pub fn set_members(origin: OriginFor<T>, members: Vec<T::AccountId>) -> DispatchResult {
+			if T::UpdateOrigin::ensure_origin(origin).is_err() {
+				return Err(BadOrigin.into())
+			}
+			Members::<T>::put(members.clone());
+			Self::deposit_event(Event::<T>::SetMembers { members });
+			Ok(())
+		}
+	}
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<T::BlockNumber> for Pallet<T> {
+		fn on_initialize(_block_number: T::BlockNumber) -> frame_support::weights::Weight {
+			let mut weight = <T as Config>::WeightInfo::on_initialize();
+			// let relaychain_block_number =
+			// 	T::RelayChainValidationDataProvider::current_block_number();
+
+			let relaychain_block_number =
+				ValidationData::<T>::get().map(|i| i.relay_parent_number).unwrap_or(0).into();
+			let mut do_on_initialize = || -> DispatchResult {
+				if !Self::is_matched() &&
+					T::ElectionSolutionStoredOffset::get()
+						.saturating_add(Self::era_start_block()) <=
+						relaychain_block_number
+				{
+					weight += <T as Config>::WeightInfo::force_matching();
+					Self::do_matching()?;
+				}
+
+				let offset = Self::offset(relaychain_block_number);
+				Self::deposit_event(Event::<T>::OnInitializeHook {
+					relay_block_number: relaychain_block_number,
+					era: offset,
+				});
+				if offset.is_zero() {
+					return Ok(())
+				}
+				weight += <T as Config>::WeightInfo::force_advance_era();
+				Self::do_advance_era(offset)
+			};
+			let _ = with_transaction(|| match do_on_initialize() {
+				Ok(()) => TransactionOutcome::Commit(Ok(())),
+				Err(err) => TransactionOutcome::Rollback(Err(err)),
+			});
+			weight
+		}
+
+		fn on_finalize(_n: T::BlockNumber) {
+			let _ = IsUpdated::<T>::clear(u32::max_value(), None);
+			if let Some(data) = T::RelayChainValidationDataProvider::validation_data() {
+				ValidationData::<T>::put(data);
+			}
+		}
+	}
+
+	impl<T: Config> Pallet<T> {
+		/// Staking pool account
+		pub fn account_id() -> T::AccountId {
+			T::PalletId::get().into_account_truncating()
+		}
+
+		/// Parachain's sovereign account
+		pub fn sovereign_account_id() -> T::AccountId {
+			T::SelfParaId::get().into_account_truncating()
+		}
+
+		/// Target era_index if users unstake in current_era
+		pub fn target_era() -> EraIndex {
+			// TODO: check if we can bond before the next era
+			// so that the one era's delay can be removed
+			Self::current_era() + T::BondingDuration::get() + 1
+		}
+
+		/// Get staking currency or return back an error
+		pub fn staking_currency() -> Result<AssetIdOf<T>, DispatchError> {
+			Self::get_staking_currency()
+				.ok_or(Error::<T>::InvalidStakingCurrency)
+				.map_err(Into::into)
+		}
+
+		/// Get liquid currency or return back an error
+		pub fn liquid_currency() -> Result<AssetIdOf<T>, DispatchError> {
```

Comment:
use AssetsRegistry. all assets must be in assets registry. similar to all other tokens.